商品購入機能

shivonnu opened the pull request:
# WHAT
商品購入機能の実装
# WHY
最終課題のフリマアプリの作成の為。

### gyazo

-  購入ボタンが出る挙動確認、購入ページ商品説明表示させた確認動画
https://gyazo.com/b89fa38ed89cde671f13c35759bf6e39**

> - 購入動画 
> https://gyazo.com/32037d61e3e881c7ae8d4328716e1a7e
> - 購入画面空白エラー表示動画
> https://gyazo.com/793ec77e70c2f91600166c68e7b22cd1
> - ターミナル・テストコード実施画像
> https://gyazo.com/090486616b1c9ca1ff92621622ce78cc

shivonnu committed:
first commit

購入と購入者情報のmodel作成とdb:migrate 購入フォームに:name等要素記入しました。

shivonnu committed:
モデル作った３つ分のコントローラー作成

コントローラーモデル１つにあたり１つ作りました。購入テーブルにpurchaseに素材ordersフォルダに入っているindexファイルを移しました。

shivonnu committed:
purchaseフォルダにcss移動コントローラーにindexメソッド追加

cssファイルをorderフォルダからpurcharseフォルダに移動しました。indexメソッドをpurchaseコントローラーに追加してページを表示できるようにしました。

shivonnu committed:
購入ページの名前をindexをshowに変更、ルーティング設定

購入ページを:id付きのページにしたかったので、ファイル名をshowに変更して、ルーティングをshowに変更しました。

shivonnu committed:
フォームに都道府県のActivehashを流用

フォームに都道府県のActivehashの要素追加、priceのREADMEのカラム名を間違ってたので修正。

shivonnu committed:
購入テーブル都道府県カラム名を変更

購入テーブルの都道府県カラム名をactivehashのshipping_area_idに統一。

shivonnu committed:
購入ページに画像と送料負担と値段表示を連動させる

@itemメソッドを用いて購入ページに該当するimageとshipping_fee_burdenとpriceを表示させる

shivonnu committed:
購入テーブルにbefore_action設置

購入テーブルにbefore_actionを設置してまとめられるように先を見越して作っておいた

shivonnu committed:
gem payjpインストールとcreateメソッド設置

gem payjpインストールとcreateメソッド設置before_actionにcreateメソッド追加

shivonnu committed:
購入フォームのバリデーション設置

購入フォーム建物名と外部キー以外のバリデーション設置

shivonnu committed:
カードのjs追加

カードのjs追加
カードフォームの挙動を確認したいので
一時的にバリデーションコメントアウト

shivonnu committed:
カードjsカード情報取得できるように変更

カーどjsのカード情報取得を確認

shivonnu committed:
カードjsAPI返信OK確認

カードPay.jp.createTokenAPI返信OK確認。

shivonnu committed:
カードparams無しでレスポンス確認

カードparams無しでレスポンス確認

shivonnu committed:
カード決済関連のフォームのid名を修正

"name"を一回消す処理が入らなかったので、idを直したら通ったので修正しました。

shivonnu committed:
formsフォルダ作成してActiveModel::Modelにより複数テーブルに保存可能にする

ActiveModel::Modelをincludeして複数のモデルを更新できるようにする。
購入してテーブルにデータが保存される、売り上げがpay.jpHPに表示されることを確認。

shivonnu committed:
環境変数定義

環境変数定義、購入確認

shivonnu committed:
環境変数一つ立て忘れ

環境変数一つしかやってなかったのでもう一つ立てる。pay_itemは可能、saveまだ出来ず。paramsは取り出せている。

shivonnu committed:
token通るようにjava変更

token通るようにjava変更

shivonnu committed:
不要バリデーション整理、DB保存確認

不要なバリデーション整理、DB保存できること確認。

shivonnu committed:
売り切れ表示の文章を追加index,show表示確認

売り切れ表示の文を追加。表示させる為にカラムを追加して結論として不要だったので削除しました。

shivonnu committed:
ページ遷移調整、編集削除ボタン購入後非表示の設定

ページ遷移調整、編集削除ボタン購入後非表示の設定、動作一通り確認。

shivonnu committed:
Bashic認証を導入

heroku pushはまだ未実行。

shivonnu committed:
tokenバリデーション、herokuの鍵環境設定

herokuの鍵環境設定。herokuにプッシュはまだ。

shivonnu committed:
テストコード記入

テストコード記入、実施確認

shivonnu committed:
郵便番号にハイフンが無いテストコード追加、全体コメントアウト空白調整

郵便番号にハイフンが無いと登録できないのを実装の条件だったので追加。コメントアウトや空白消す。テストコードとrubocop実行確認。

masahirokurauchi commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,44 @@
+     @purchases = Purchase.all
+     if    current_user.id == @item.user_id
+           redirect_to root_path
+     elsif @purchases.where(item_id: @item.id).present?
```
＠itemに紐づくpurchaseレコードが存在しなければ、
という条件にしましょう

masahirokurauchi commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,44 @@
+ 
+  def index
+     @purchase = PurchaseShippingAddress.new
+     @purchases = Purchase.all
```
指摘させていただきました条件で実装できれば、こちらは必要ありません

masahirokurauchi commented on app/models/purchase_shipping_address.rb:
```diff
@@ -0,0 +1,20 @@
+  include ActiveModel::Model
+  attr_accessor  :postal_code, :shipping_area_id, :city, :address, :building_name, :phone_number, :user_id, :item_id, :token, :purchase_id
+
+  validates  :token,            presence: true
+  validates  :shipping_area_id, presence: true
+  validates  :city,             presence: true
+  validates  :address,          presence: true
+  validates  :phone_number,     presence: true
+  with_options presence: true do
+    validates :postal_code, format: {with: /\A[0-9]{3}-[0-9]{4}\z/, message: "is invalid. Include hyphen(-)."}
+    validates :phone_number, format: {with: /\A[0-9]{9,11}\z/, message: "is invalid. Within 11 digits. Do not put hyphens."}
+  end
```
"バリデーションは「with_options」を使用することで、共通したオプションを付けることが可能となります。
こちらのカリキュラムの内容を参考にまとめてみてはいかがでしょう？
https://master.tech-camp.in/curriculums/5001 "

masahirokurauchi reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

shivonnu committed:
購入済判断のpurchase.whereをitemテーブルに紐づく表示に修正

購入済ページ変遷出来ないようにするの部分でご指摘頂きましたので、トップページindexと商品詳細の売り切れ表示表記部分も合わせて修正しました。

shivonnu committed:
商品詳細ページ２箇所purchase.where修正し忘れ修正、バリデーションwith_optionsの中に３テーブル分まとめ

商品詳細ページローカル環境のページでエラー確認して漏れに気づいたので修正、バリデーションwith_optionsの中に３テーブル分まとめました。

shivonnu commented:
コメントありがとうございます！
修正させて頂きましたので再度コードレビュー依頼させて頂きます！

yupimaru commented on app/javascript/card.js:
```diff
@@ -0,0 +1,37 @@
+    };
+    console.log(card)
+    Payjp.createToken(card, (status, response) => {
+      console.log(status)
```
```js
console.log(card)
console.log(status)
```
17行目、19行目に記載されている`console.log()`ですが、こちらはデバッグ用コードとなります。
デバッグが終了した後は、削除しておきましょう。

プログラム実行においては不要なコードであるということ、加えてJavascriptでの`console.log()`はブラウザで検証ツールを開いた際に出力されてしまうため、意図しないデータが第三者に漏れてしまう可能性があるためです。

yupimaru commented on app/views/items/show.html.erb:
```diff
@@ -25,15 +24,17 @@
     
   <% if user_signed_in?%>
     <% if user_signed_in? && current_user.id == @item.user_id%>
```
```erb
<% if user_signed_in?%>
  <% if user_signed_in? && current_user.id == @item.user_id%>
```
25・26行目において、`user_signed_in?`が重複しているように見受けられます。
以下のように記述しても、同様の結果が得られそうです。

```erb
<% if user_signed_in?%>
  <% if current_user.id == @item.user_id%>
```

yupimaru commented on app/views/purchases/index.html.erb:
```diff
@@ -1,38 +1,36 @@
+      <%= image_tag @item.image, class: 'buy-item-img' if @item.image.attached?%>
       <div class='buy-item-right-content'>
         <h2 class='buy-item-text'>
           <%= "商品説明" %>
```
```erb
<%= "商品説明" %>
```

こちらはDB上に保存されている**商品説明**が出力されるように修正しましょう。

yupimaru reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

shivonnu committed:
consol.log消去修正、if user_signed_in?二重の修正

consol.log残っていると不要なのと情報漏れる可能性があるので消去しました。if user_signed_in?二重になっているので修正して動作確認しました。コードの並びがガタガタで見にくかったので修正しました。

shivonnu committed:
購入ページの商品説明が表示されるように修正

購入ページの商品説明が表示されるように修正しました。

shivonnu commented:
コメントありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

chirikuzu commented on config/routes.rb:
```diff
@@ -1,6 +1,10 @@
   devise_for :users
-  resources :items
+  resources :items do
+    resources :purchases
```
こちら`resources :purchases`にはonlyでindexとcreateのみ設定してもいいかもしれません。
不要なルーティングはアプリケーション全体の処理に影響します。

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
細かいですが追加でコメントいたしましたので、ご確認お願いします！

shivonnu committed:
購入テーブルにindex,createだけのルーティング追加

処理を安定させる為に購入テーブルのルーティングをindex,createだけにしました。itemsは全種類実行したいのでそのままにしました。購入者情報は不要なルーティングだったので削除しました。コメントアウトも削除しました。

shivonnu commented:
コメントありがとうございます！
修正しましたので再度コードレビュー依頼させて頂きます！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
引き続き実装頑張っていきましょう！

shivonnu commented:
ありがとうございます！引き続き頑張ります！
